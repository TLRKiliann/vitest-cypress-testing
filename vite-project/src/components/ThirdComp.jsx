import { useState, useRef, useEffect } from 'react'

const ThirdComp = ({id, name, inputValue, dis, display, setDisplay}) => {

  const [editBool, setEditBool] = useState(false)
  const [edit, setEdit] = useState(name)

  const myRef = useRef()

  const handleEdit = (e, id) => {
    e.preventDefault()
    setDisplay(display?.map((dis) => dis.id === id 
      ? {...dis, name: edit, editor: true}
      : dis ))
    setEditBool((editBool) => !editBool)
  }

  const handleDelete = (id) => {
    const delId = display.filter((d) => d.id !== id)
    setDisplay(delId)
  }

  const handleEventChangeEdit = (e) => {
    setEdit(e.target.value);
  }

  useEffect(() => {
    myRef?.current?.focus();
  }, [edit])

  return (
    <>
      <div data-testid="testid-divthird">
        {editBool === true && dis?.editor === true ? (
          <input
            data-testid="testid-eventedit"
            ref={myRef}
            value={edit} 
            onChange={(e) => handleEventChangeEdit(e)}
          />
          ) : (
          <div className="div--li">
            <p data-testid="testid-datestr">
              <li>
                {dis?.date?.toDateString()} - {name}
              </li>
            </p>
          </div>
        )}      
        <button 
          data-testid="testid-edit"
          type="button" 
          onClick={(e) => handleEdit(e, dis?.id)}
        >
          {editBool === false ? "Edit" : "Register"}
        </button>
        <button
          data-testid="testid-del"
          type="button"
          onClick={() => handleDelete(dis?.id)}
        >
          Delete
        </button>
      </div>
    </>
  )
}
export default ThirdComp;