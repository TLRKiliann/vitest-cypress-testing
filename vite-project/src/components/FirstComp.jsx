import { useState, useRef, useEffect } from 'react'
import ThirdComp from './ThirdComp'

let countId = 0;

export default function FirstComp({showing, setShowing}) {
  
  const [inputValue, setInputValue] = useState("")
  const [display, setDisplay] = useState([])
  const date = new Date()

  const handleSubmit = (e) => {
    e.preventDefault()
    if (inputValue) {
      setDisplay([...display, {date, id: countId++, name: inputValue, editor: false}])
    }
    setInputValue("")
  }

  const handleEventChange = (e) => {
    setInputValue(e.target.value)
  }

  const handleReverse = () => {
    const reversed = display.reverse()
    setDisplay([...reversed])
  }

  return (
    <>
      <h1 
        data-testid="first-comp"
        style={showing ? {color: "cyan"} : {color: "hotpink"}}>
        React-Dev
      </h1>

      <form data-testid="testid-form" onSubmit={handleSubmit}>
        <input
          data-testid="testid-event"
          type="text"
          value={inputValue}
          onChange={(e) => handleEventChange(e)}
        />
        <button
          data-testid="testid"
          type="submit">
          Enter
        </button>
      </form>

      <button
        data-testid="testid-reverse"
        type="button"
        onClick={handleReverse}>
        Reverse
      </button>

      {display.map((dis) => (
        <ThirdComp
          key={dis.id}
          inputValue={inputValue}
          dis={dis}
          name={dis.name}
          editor={dis.editor}
          display={display}
          setDisplay={setDisplay}
        />
      ))}
    </>
  )
}