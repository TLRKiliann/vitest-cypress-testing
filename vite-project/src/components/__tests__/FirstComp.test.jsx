import React from 'react'
import { test, expect } from 'vitest'
import { fireEvent, screen, render } from '@testing-library/react'
import { create } from "react-test-renderer"
import "@testing-library/jest-dom"
import FirstComp from '../FirstComp.jsx'
import ThirdComp from '../FirstComp.jsx'
import { handleSubmit } from '../FirstComp.jsx'
import { handleEventChange } from '../FirstComp.jsx'
import { handleReverse } from '../FirstComp.jsx'

/*
test("mocking useState", () => {
  vi.mock("../ThirdComp.jsx", () => {
    const actual = vi.importActual("../ThirdComp.jsx")
    return actual;
  })
});
*/

test('MatchSnapShot FirstComp', () => {
  const tree = create(
  	<FirstComp />
  )
  expect(tree.toJSON()).toMatchSnapshot()
});

describe("testing children FirstComp", () => {
  test('testing function handleSubmit', () => {
    const beverage = {...["abc"], id: 2, name: "Doug", editor: false }
    const handleSubmit = vi.fn(beverage => beverage)
    handleSubmit(beverage)
    render(
      <FirstComp>
        <button data-testid="testid">
          Enter
        </button>
      </FirstComp>
    )
    fireEvent.submit(screen.getByTestId('testid'))
    expect(handleSubmit).toHaveReturnedWith({ 
      ...['abc'], id: 2, name: "Doug", editor: false 
    })
    expect(handleSubmit).toHaveBeenCalledTimes(1)
  })
});

//Submit Form
test("submit form test", () => {
  const beverage = {id: 3, name: "doug", editor: false};
  const handleSubmit = vi.fn(beverage => beverage.id);
  handleSubmit(beverage);
  render(
    <FirstComp>
      <form 
        data-testid="testid-form"
        onSubmit={handleSubmit}
      >  
      </form>
    </FirstComp>
  );
  fireEvent.submit(screen.getByTestId("testid-form"));
  expect(handleSubmit).toHaveReturnedWith(3);
  expect(handleSubmit).toHaveBeenCalledTimes(1);
});

describe("testing children FirstComp", () => {
  test('testing function handleReverse', () => {
    const beverage = {id: 2, id: 3}
    const handleReverse = vi.fn(beverage => beverage.id)
    handleReverse(beverage)
    render(
      <FirstComp>
        <button data-testid="testid-reverse">
          Reverse
        </button>
      </FirstComp>
    )
    fireEvent.click(screen.getByTestId('testid-reverse'))
    expect(handleReverse).toHaveReturnedWith(3, 2)
    expect(handleReverse).toHaveBeenCalledTimes(1)
    expect(handleReverse).toBeDefined();
  })
});

describe("captures event from FirstComp", async () => {
  test('testing function handleEventChange', () => {
    function handleEventChange(evt) {
      expect(evt.target.value).toEqual("changedvalue");
    }
    render(
      <FirstComp>
        <input
          data-testid="testid-event"
          type="text"
          value=""
          onChange={(evt) => handleEventChange(evt)}
        />
      </FirstComp>
    )
    const node = screen.getByTestId("testid-event");
    fireEvent.change(node, { target: { value: "changedvalue" }});
    expect(handleEventChange).toBeDefined();
  })
});

test("captures nbr of btn", () => {
  const { getByRole } = render(
    <FirstComp />
  );
  const button = screen.getAllByRole('button');
  expect(button).toHaveLength(2);
});

test('handleSubmit updates display state and clears inputValue', () => {
  const setDisplay = vi.fn();
  
  const { getByTestId } = render(<FirstComp setDisplay={setDisplay} />);
  
  const form = getByTestId('testid-form');
  const input = getByTestId('testid-event');
  
  // Simuler une saisie dans le champ de texte
  fireEvent.change(input, { target: { value: 'Test input' } });
  
  // Soumettre le formulaire
  fireEvent.submit(form);
  
  // Vérifier si setDisplay a été appelé avec les bons arguments
  
  // Vérifier si l'état inputValue a été effacé
  expect(input.value).toBe('');
});


test('renders component with "cyan" color when showing is true', () => {
  render(
    <FirstComp showing={true} setShowing={vi.fn()} />
  );
  const heading = screen.getByTestId('first-comp');
  expect(heading).toHaveStyle({ color: 'cyan' });
});

test('renders component with "hotpink" color when showing is false', () => {
  render(
    <FirstComp showing={false} setShowing={vi.fn()} />
  );
  const heading = screen.getByTestId('first-comp');
  expect(heading).toHaveStyle({ color: 'hotpink' });
});

/*
test("getByTestId will throw with multiple testIDs", () => {
  const {getAllByTestId, getByTestId} = render(
    <ThirdComp>
      <input data-testid="testid-datestr">a</input>
      <input data-testid="testid-datestr">b</input>
    </ThirdComp>
  );
  expect(getAllByTestId("testid-eventedit")).toHaveLength(2); // OK
  getByTestId("testid-datestr"); // => Error: Found multiple elements with testID: foo
});

test("should find text content in all children", () => {
  const {getAllByTestId} = render(
    <div>
      {[..."abcd"].map((e, i) => 
        <input key={e + i} data-testid="testid-datestr">{e}</input>
      )}
    </div>
  );
  expect(getAllByTestId("testid-datestr")).toHaveLength(4);
  [..."abcd"].forEach((e, i) => {
    expect(getAllByTestId("testid-datestr")[i].textContent).toEqual(e);
  });
  // or:
  //const contents = getAllByTestId("foo").map(e => e.textContent);
  //expect(contents).toEqual([..."abcd"]);
});
*/