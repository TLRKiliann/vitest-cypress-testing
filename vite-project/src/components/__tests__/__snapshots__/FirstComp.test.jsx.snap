// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MatchSnapShot FirstComp 1`] = `
[
  <h1
    data-testid="first-comp"
    style={
      {
        "color": "hotpink",
      }
    }
  >
    React-Dev
  </h1>,
  <form
    data-testid="testid-form"
    onSubmit={[Function]}
  >
    <input
      data-testid="testid-event"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      data-testid="testid"
      type="submit"
    >
      Enter
    </button>
  </form>,
  <button
    data-testid="testid-reverse"
    onClick={[Function]}
    type="button"
  >
    Reverse
  </button>,
]
`;
